processing/src/main/java/io/druid/query/metadata/SegmentAnalyzer.java

```diff
@@ -82,7 +88,42 @@
       columns.put(columnName, analysis);
     }
 
-    columns.put(Column.TIME_COLUMN_NAME, lengthBasedAnalysis(index.getColumn(Column.TIME_COLUMN_NAME), NUM_BYTES_IN_TIMESTAMP));
+    columns.put(
+        Column.TIME_COLUMN_NAME,
+        lengthBasedAnalysis(index.getColumn(Column.TIME_COLUMN_NAME), NUM_BYTES_IN_TIMESTAMP)
+    );
+
+    return columns;
+  }
+
+  public Map<String, ColumnAnalysis> analyzeStorageAdapter(StorageAdapter adapter)
+  {
+    Preconditions.checkNotNull(adapter, "Adapter cannot be null");
+    Map<String, ColumnAnalysis> columns = Maps.newTreeMap();
+    List<String> columnNames = getStorageAdapterColumnNames(adapter);
+
+    for (String columnName : columnNames) {
+      final ColumnCapabilities capabilities = adapter.getColumnCapabilities(columnName);
+
+      ValueType capType = capabilities.getType();
+      int cardinality = adapter.getDimensionCardinality(columnName);
+      ColumnAnalysis analysis = new ColumnAnalysis(
+          capType.name(),
+          0,
+          capType == ValueType.STRING ? cardinality : null,
+          null
+      );
+      columns.put(columnName, analysis);
+    }
+
+    /*
+     * Time column is not available from IncrementalIndex.
+     * Size information is not calculated here.
+    */
+    columns.put(
+        Column.TIME_COLUMN_NAME,
+        new ColumnAnalysis(ValueType.LONG.name(), 0, null, null)
+    );
```

Comment:
This could be a single line // comment, also imo the real issue is that StorageAdapters don't tell you how many rows they have. (you don't actually need to read the time column to know how big it is; you just have to know how many rows the segment has)